modeling/random_forest.py

```diff
@@ -0,0 +1,206 @@
+                    verbose = 2)
+    search.fit(X_train, y_train[label])
+    model = search.best_estimator_
+    y_pred_val = model.predict(X_val)
+    y_pred_train = model.predict(X_train)
+    y_pred_val_proba = model.predict_proba(X_val)[:,1]
+    y_pred_train_proba = model.predict_proba(X_train)[:,1]
+    best_params=search.best_params_
+    if best_params['vectorizer__stop_words']!=None:
+        best_params['vectorizer__stop_words']='NLTK-German'
+    return model,best_params, pd.Series(y_pred_val, name=label), pd.Series(y_pred_train, name=label), pd.Series(y_pred_val_proba, name=label), pd.Series(y_pred_train_proba)
+
+
+def get_params(label, best_params):
+    params = {
+    "model": "random_forest_cv_testrun",
+    "label": label,
+    "vectorizer": "tfidf",
+    "normalization": "lower",
+    "best_params": best_params
+    }
+    return params
+
+
+def best_th(y_pred_proba, y_true, label):
+        best_th = 0.0
+        best_f1 = 0.0
+        for th in np.arange(0.05, 0.96, 0.05):
+            y_pred_temp = y_pred_proba.apply(lambda x: 1 if x>=th else 0)
+            if f1_score(y_true[label], y_pred_temp)>best_th:
+                best_th = th
+                best_f1 = f1_score(y_true[label], y_pred_temp)
+        return best_th
+
+
+def log_metrics(f1_val, recall_val, precision_val, f1_train, recall_train, precision_train):
+    mlflow.log_metric("val - " + "F1", f1_val)
+    mlflow.log_metric("val - " + "recall", recall_val)
+    mlflow.log_metric("val - " + "precision", precision_val)
+    mlflow.log_metric("train - " + "F1", f1_train)
+    mlflow.log_metric("train - " + "recall", recall_train)
+    mlflow.log_metric("train - " + "precision", precision_train)
+
+
+def sv_model(params, model):
+    mlflow.log_params(params)
+    mlflow.set_tag("running_from_jupyter", "True")
+    #mlflow.log_artifact("./models")
+    #mlflow.sklearn.log_model(model, "model")
+    path = f"models/{params['model']+params['label']}_{datetime.now().strftime('%Y-%m-%d_%H%M%S')}"
+    save_model(sk_model=model, path=path)
+
+
+def endrun():
+    mlflow.end_run()
+
+
+# +
+def train_rf():
+    data_train, data_val = get_data()
+    data_train=clean_data(data_train)
+    data_val=clean_data(data_val)
+    
+    y_cols=['label_argumentsused', 'label_discriminating', 'label_inappropriate',
+       'label_offtopic', 'label_personalstories', 'label_possiblyfeedback',
+       'label_sentimentnegative',
+        'label_sentimentneutral',
+       'label_sentimentpositive']
+    
+    X_train, y_train = split_Xy(data_train, y_cols)
+    X_val, y_val = split_Xy(data_val, y_cols)
+    
+    pipe, grid = get_pipe_grid()
+    grid_mlflow = {
+    'vec_ngram_range' : [(1,1), (1,2), (1,3)],
+    'vec_stop_words' : ['NLTK-German', None],
+    'vec_max_df': [0.5, 1.0],
+    'vec_min_df': [0, 5],
+    'cl_n_estimators': [100, 1000], 
+    'cl_max_depth': [1, 5, 10],
+    'cl_min_samples_leaf': [5, 10, 40]
+    }
+    
+    
+    for label in y_cols:
+        endrun()
```

Comment:
One note about `start_mlflow()` and `end_mlflow()`:
If you removed the `start_run`-lines and just set the MLFlow params in `start_mlflow()` you could use Python's `with <context>` syntax here:
```python
...
for label in y_cols:
    set_mlflow_params() # where you set the TRACKING_URI and EXPERIMENT_NAME
    with mlflow.start_run():
        # here you put all the code that is in your original for-label-loop
```
What is the advantage of the `with <context>`? It closes the run when it is interrupted, e.g. due to errors.
Why did we do it differently in the original notebook? Because we had the part in the for-label-loop spread across different cells and therefore couldn't put it into one `with <context>`-statement.